source/visionEnhancementProviders/screenCurtain.py

```diff
@@ -66,30 +86,247 @@ class Magnification:
+
+	def _onDialogActivated(self, evt):
+		# focus is normally set to the first child, however, we want people to easily be able to cancel this
+		# dialog
+		super()._onDialogActivated(evt)
+		self.noButton.SetFocus()
+
+	def _onShowEvt(self, evt):
+		"""When no other dialogs have been opened first, focus lands in the wrong place (on the checkbox),
+		so we correct it after the dialog is opened.
+		"""
+		if evt.IsShown():
+			self.noButton.SetFocus()
+		super()._onShowEvt(evt)
+
+
+class ScreenCurtainGuiPanel(
+		gui.DriverSettingsMixin,
+		gui.SettingsPanel,
+):
+
+	_enabledCheckbox: wx.CheckBox
+	_enableCheckSizer: wx.BoxSizer
+
+	from gui.settingsDialogs import VisionProviderStateControl
+
+	def __init__(
+			self,
+			parent,
+			providerControl: VisionProviderStateControl
+	):
+		self._providerControl = providerControl
+		super().__init__(parent)
+
+	def _buildGui(self):
+		self.mainSizer = wx.BoxSizer(wx.VERTICAL)
+
+		self._enabledCheckbox = wx.CheckBox(
+			self,
+			#  Translators: option to enable screen curtain in the vision settings panel
+			label=_("Make screen black (immediate effect)")
+		)
+
+		self.mainSizer.Add(self._enabledCheckbox)
+		self.mainSizer.AddSpacer(size=self.scaleSize(10))
+		# this options separator is done with text rather than a group box because a groupbox is too verbose,
+		# but visually some separation is helpful, since the rest of the options are really sub-settings.
+		self.optionsText = wx.StaticText(
+			self,
+			# Translators: The label for a group box containing the NVDA highlighter options.
+			label=_("Options:")
+		)
+		self.mainSizer.Add(self.optionsText)
+		self.lastControl = self.optionsText
+		self.settingsSizer = wx.BoxSizer(wx.VERTICAL)
+		self.makeSettings(self.settingsSizer)
+		self.mainSizer.Add(self.settingsSizer, border=self.scaleSize(15), flag=wx.LEFT | wx.EXPAND)
+		self.mainSizer.Fit(self)
+		self.SetSizer(self.mainSizer)
+
+	def getSettings(self) -> ScreenCurtainSettings:
+		return ScreenCurtainProvider.getSettings()
+
+	def makeSettings(self, sizer: wx.BoxSizer):
+		self.updateDriverSettings()
+		self.Bind(wx.EVT_CHECKBOX, self._onCheckEvent)
+
+	def onPanelActivated(self):
+		self.lastControl = self._enabledCheckbox
+
+	def _onCheckEvent(self, evt: wx.CommandEvent):
+		if evt.GetEventObject() is self._enabledCheckbox:
+			self._ensureEnableState(evt.IsChecked())
+
+	def _ensureEnableState(self, shouldBeEnabled: bool):
+		currentlyEnabled = bool(self._providerControl.getProviderInstance())
+		if shouldBeEnabled and not currentlyEnabled:
+			confirmed = self.confirmInitWithUser()
+			if confirmed:
+				self._providerControl.startProvider()
+			else:
+				self._enabledCheckbox.SetValue(False)
+		elif not shouldBeEnabled and currentlyEnabled:
+			self._providerControl.terminateProvider()
+
+	def confirmInitWithUser(self) -> bool:
+		settingsStorage = self._getSettingsStorage()
+		if not settingsStorage.warnOnLoad:
+			return True
+		parent = self
+		dlg = WarnOnLoadDialog(

```

Comment:
Could you put this into a with statement? This ensures that the dlg is destroyed when leaving the with statement.

Reply:
I wondered about this, it probably doesn't hurt, and may reduce the chance of a bug, but is there a reason this must happen? Will it currently cause a bug I'm not thinking of?

Reply:
I don't think it will hurt, it might take some time for the garbage collector to kick in and destroy the dialog.

Note that you can still return from a with statement